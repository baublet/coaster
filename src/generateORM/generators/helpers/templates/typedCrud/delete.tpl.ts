/** Deletes a $$entityName */
export async function delete$$entityName(
  entity: $$entityName,
  connection: ConnectionOrTransaction
): Promise<number> {
  return $$rawBaseQueryFunctionName<number>(connection)
    .where("$$tablePrimaryKeyColumn", "=", entity.$$tablePrimaryKeyColumn)
    .delete()
    .limit(1);
}

/** Delete one or more $$pluralEntityName under specific conditions */
export async function delete$$entityNameWhere(
  query: (query: knex.QueryBuilder<$$rawEntityTypeName, number>) => unknown,
  connection: ConnectionOrTransaction
): Promise<number> {
  const queryBuilder = $$rawBaseQueryFunctionName<number>(connection);
  await query(queryBuilder);
  queryBuilder.delete();
  return queryBuilder;
}
