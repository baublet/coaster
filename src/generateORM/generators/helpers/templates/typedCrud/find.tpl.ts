/** Find many $$pluralEntityName in the database by a constraint function */
export async function find$$pluralEntityName(
  query: (
    query: Knex.QueryBuilder<$$rawEntityTypeName, $$rawEntityTypeName[]>
  ) => unknown,
  connection: ConnectionOrTransaction
): Promise<$$entityName[]> {
  const queryBuilder = $$rawBaseQueryFunctionName(connection);
  await query(queryBuilder);
  const results = await queryBuilder;
  return results.map((rawEntity) => $$rawToNamedFunctionName(rawEntity));
}

/** Find a $$entityName in the database or fail */
export async function find$$entityNameOrFail(
  query: (
    query: Knex.QueryBuilder<$$rawEntityTypeName, $$rawEntityTypeName[]>
  ) => unknown,
  connection: ConnectionOrTransaction
): Promise<$$entityName> {
  const queryBuilder = $$rawBaseQueryFunctionName(connection);
  await query(queryBuilder);
  queryBuilder.limit(1);
  const results = await queryBuilder;
  if (results.length === 0) {
    throw new Error(
      "Error! Unable to find $$entityName in find$$entityNameOrFail call"
    );
  }
  return $$rawToNamedFunctionName(results[0]);
}
