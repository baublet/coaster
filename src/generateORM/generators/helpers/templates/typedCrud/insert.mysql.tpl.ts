/**
 * Insert a single $$entityName into the database, returning the inserted ID
 */
export async function $$insertSingleFunctionName(
  input: $$entityInputType,
  connection: ConnectionOrTransaction
): Promise<$$returnType> {
  const rawInput = $$namedToRawFunctionName(input);
  const result = await $$rawBaseQueryFunctionName(connection).insert(rawInput);
  return $$rawToNamedFunctionName(result[0]);
}

/**
 * Inserts one ore more $$pluralEntityName into the database, returning the inserted IDs
 */
export async function $$insertPluralFunctionName(
  input: $$entityInputType[],
  connection: ConnectionOrTransaction
): Promise<$$returnType[]> {
  const rawInput = input.map((input) => $$namedToRawFunctionName(input));
  const results = await $$rawBaseQueryFunctionName(connection).insert(rawInput);
  return results.map((rawEntity) => $$rawToNamedFunctionName(rawEntity));
}
