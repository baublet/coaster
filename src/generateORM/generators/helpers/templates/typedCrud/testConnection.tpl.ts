const connection = knex($$testConnectionString);

const testMigrations: ((...args: any[]) => Promise<void>)[] = [];

beforeAll(async () => {
  for(const migration of testMigrations) {
    await connection.transaction(trx => migration(trx));
  }
});

afterAll(async () => {
  await connection.destroy();
});
