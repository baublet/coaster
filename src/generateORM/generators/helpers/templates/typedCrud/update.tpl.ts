/** Update a single $$entityName. Resolves the number of affected rows. */
export function $$updateSingleFunctionName(
  entity: $$entityInputType,
  connection: ConnectionOrTransaction
): Promise<number> {
  const { $$tablePrimaryKeyColumn, ...rawInput } =
    $$namedToRawFunctionName(entity);
  return $$rawBaseQueryFunctionName(connection)
    .update(rawInput)
    .where("$$tablePrimaryKeyColumn", "=", entity.$$tablePrimaryKeyColumn)
    .limit(1);
}

/** Update one or more $$pluralEntityName under specific conditions. Resolves the number of affected rows. */
export async function $$updatePluralFunctionName(
  updatePayload: Partial<$$entityInputType>,
  query: (query: Knex.QueryBuilder<$$rawEntityTypeName, number>) => unknown,
  connection: ConnectionOrTransaction
): Promise<number> {
  const rawUpdatePayload = $$namedToRawFunctionName(updatePayload);
  const queryBuilder =
    $$rawBaseQueryFunctionName(connection).update(rawUpdatePayload);
  await query(queryBuilder);
  return await queryBuilder.clone();
}
