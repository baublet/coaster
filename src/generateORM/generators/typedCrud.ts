import { MetaData, GeneratorResult } from ".";
import { generateNames } from "../../generateNames";
import { RawSchema } from "../drivers";
import { getSchemaAndTablePath } from "./helpers";

/**
 * Generates the lowest-level possible accessors and mutators for data in a
 * table.
 */
export const typedCrud = (
  schema: RawSchema,
  metaData: MetaData
): GeneratorResult => {
  metaData.setHeader(
    "knex",
    metaData.templateManager.render({ template: "rawBaseQuery/knex" })
  );

  let code = "";
  const testCode = "";

  for (const table of schema.tables) {
    const schemaAndTablePath = getSchemaAndTablePath(schema.name, table.name);
    const entityName = metaData.tableEntityNames.get(schemaAndTablePath);
    const pluralEntityName = generateNames(entityName).pluralPascal;
    const rawBaseQueryFunctionName =
      metaData.rawBaseQueryFunctionNames.get(schemaAndTablePath);
    const rawEntityTypeName =
      metaData.tableRawEntityNames.get(schemaAndTablePath);
    const rawToNamedFunctionName =
      metaData.transformerFunctionNames[rawEntityTypeName][entityName];
    const namedToRawFunctionName =
      metaData.transformerFunctionNames[entityName][rawEntityTypeName];
    const entityInputType =
      metaData.namedEntityInputTypeNames.get(schemaAndTablePath);

    // Create

    if (schema.flavor === "mysql") {
      code += metaData.templateManager.render({
        template: "typedCrud/insert.mysql",
        variables: {
          entityInputType,
          entityName,
          pluralEntityName,
          returnType: table.primaryKeyType,
          rawToNamedFunctionName,
          rawBaseQueryFunctionName,
        },
      });
    } else {
      code += metaData.templateManager.render({
        template: "typedCrud/insert.pg",
        variables: {
          entityInputType,
          entityName,
          pluralEntityName,
          rawToNamedFunctionName,
          rawBaseQueryFunctionName,
        },
      });
    }

    // Read
    code += metaData.templateManager.render({
      template: "typedCrud/find",
      variables: {
        entityName,
        pluralEntityName,
        rawBaseQueryFunctionName,
        rawEntityTypeName,
        rawToNamedFunctionName,
      },
    });

    // Update
    code += metaData.templateManager.render({
      template: "typedCrud/update",
      variables: {
        entityInputType,
        namedToRawFunctionName,
        rawBaseQueryFunctionName,
        rawEntityTypeName,
        tablePrimaryKeyColumn: table.primaryKeyType,
        pluralEntityName,
        entityName,
      },
    });

    // Delete
    code += `/** Deletes a ${entityName} */\n`;
    code += `export async function delete${entityName}(\n`;
    code += `  entity: ${entityName},\n`;
    code += `  connection: ConnectionOrTransaction\n`;
    code += `): Promise<number> {\n`;
    code += `  return ${rawBaseQueryFunctionName}<number>(connection)\n`;
    code += `    .where("${table.primaryKeyColumn}", "=", entity.${table.primaryKeyColumn})\n`;
    code += `    .delete()\n`;
    code += `    .limit(1);\n`;
    code += `}\n\n`;

    // Delete where
    code += `/** Delete one or more ${pluralEntityName} under specific conditions */\n`;
    code += `export async function delete${entityName}Where(\n`;
    code += `  query: (query: knex.QueryBuilder<${rawEntityTypeName}, number>) => unknown,\n`;
    code += `  connection: ConnectionOrTransaction\n`;
    code += `): Promise<number> {\n`;
    code += `  const queryBuilder = ${rawBaseQueryFunctionName}<number>(connection);\n`;
    code += `  await query(queryBuilder);\n`;
    code += `  queryBuilder.delete();\n`;
    code += `  return queryBuilder;\n`;
    code += `}\n\n`;
  }

  return {
    code,
    testCode,
  };
};
