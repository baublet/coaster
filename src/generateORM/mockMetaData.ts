import { MetaData } from "./generators";
import { getTemplateManager } from "../generateORM/generators/helpers";

export function getMockMetaData(): MetaData {
  return {
    testConnectionVariable: "connection",
    codeOutputFullPath: "./generated.ts",
    generateTestCode: false,
    entityTableNames: new Map(),
    setHeader: jest.fn(),
    setTestHeader: jest.fn(),
    tableEntityNames: new Map(),
    rawBaseQueryFunctionNames: new Map(),
    transformerFunctionNames: {},
    typeAssertionFunctionNames: new Map(),
    typeGuardFunctionNames: new Map(),
    tableRawEntityNames: new Map(),
    namedEntityColumnNames: new Map(),
    namedEntityInputTypeNames: new Map(),
    namedDatabaseEnumNames: new Map(),
    rawDatabaseEnumNames: new Map(),
    namedCreateTestEntityFunctionNames: new Map(),
    rawCreateTestEntityFunctionNames: new Map(),
    rawEnumValues: new Map(),
    templateManager: getTemplateManager(),
    namedMockEntityFunctionNames: new Map(),
    rawMockEntityFunctionNames: new Map(),
  };
}
