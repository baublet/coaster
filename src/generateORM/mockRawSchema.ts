import { RawSchema } from "./drivers";

export const getMockRawSchema: () => RawSchema = () => ({
  name: "public",
  flavor: "pg",
  enums: [
    {
      name: "user_account_source",
      values: ["github", "bitbucket"],
    },
  ],
  tables: [
    {
      name: "user",
      primaryKeyColumn: "id",
      primaryKeyType: "string",
      uniqueConstraints: [],
      columns: [
        {
          name: "id",
          columnType: "columnType",
          nullable: false,
          type: "string",
          foreignKeys: [],
          hasDefault: true,
        },
        {
          name: "name",
          comment: "Users don't have to provide their name",
          columnType: "columnType",
          nullable: true,
          type: "string",
          foreignKeys: [],
          hasDefault: false,
        },
        {
          name: "data",
          columnType: "columnType",
          nullable: false,
          type: "AnyJson",
          foreignKeys: [],
          hasDefault: false,
        },
      ],
    },
    {
      name: "user_account",
      comment: "Singular users can have multiple accounts",
      primaryKeyColumn: "id",
      primaryKeyType: "string",
      uniqueConstraints: [],
      columns: [
        {
          name: "id",
          columnType: "columnType",
          nullable: false,
          type: "unknown",
          foreignKeys: [],
          hasDefault: true,
        },
        {
          name: "source",
          comment: "Who sent us this users' data?",
          columnType: "columnType",
          nullable: false,
          type: "enum",
          foreignKeys: [],
          hasDefault: false,
          enumPath: "public.user_account_source",
        },
        {
          name: "source_data",
          comment: "The raw data sent from the source provider",
          columnType: "columnType",
          nullable: false,
          type: "AnyJson",
          foreignKeys: [],
          hasDefault: false,
        },
      ],
    },
  ],
});
