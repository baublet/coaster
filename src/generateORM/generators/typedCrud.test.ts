import { typedCrud } from "./typedCrud";

import { getMockRawSchema } from "../mockRawSchema";
import { getMockMetaData } from "../mockMetaData";

const metaData = getMockMetaData();
const mockSchema = getMockRawSchema();

metaData.tableRawEntityNames.set("public.user", "RawUser");
metaData.tableRawEntityNames.set("public.user_account", "RawUserAccount");
metaData.tableEntityNames.set("public.user", "User");
metaData.tableEntityNames.set("public.user_account", "UserAccount");

metaData.rawBaseQueryFunctionNames.set("public.user", "RawUsers");
metaData.rawBaseQueryFunctionNames.set(
  "public.user_account",
  "RawUserAccounts"
);

metaData.transformerFunctionNames["RawUser"] = {};
metaData.transformerFunctionNames["RawUser"]["User"] = "rawUserToUser";
metaData.transformerFunctionNames["RawUserAccount"] = {};
metaData.transformerFunctionNames["RawUserAccount"]["UserAccount"] =
  "rawUserAccountToUserAccount";

metaData.transformerFunctionNames["User"] = {};
metaData.transformerFunctionNames["User"]["RawUser"] = "userToRawUser";
metaData.transformerFunctionNames["UserAccount"] = {};
metaData.transformerFunctionNames["UserAccount"]["RawUserAccount"] =
  "userAccountToRawUserAccount";

metaData.namedEntityInputTypeNames.set("public.user", "UserInput");
metaData.namedEntityInputTypeNames.set(
  "public.user_account",
  "UserAccountInput"
);

it("generates some code: PG flavored", () => {
  expect(typedCrud(mockSchema, metaData)).toMatchSnapshot();
});

it("generates some code: MySQL flavored", () => {
  const mockSchema = getMockRawSchema();
  mockSchema.flavor = "mysql";

  expect(typedCrud(mockSchema, metaData)).toMatchSnapshot();
});
