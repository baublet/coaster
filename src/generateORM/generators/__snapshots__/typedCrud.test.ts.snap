// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates some code: MySQL flavored 1`] = `
Object {
  "code": "/**
 * Insert a single User into the database, returning the inserted ID
 */
export async function insertUser(
  input: UserInput,
  connection: ConnectionOrTransaction
): Promise<string> {
  const rawInput = userToRawUser(input);
  const result = await RawUsers(connection).insert(rawInput);
  return rawUserToUser(result[0]);
}

/**
 * Inserts one ore more Users into the database, returning the inserted IDs
 */
export async function insertUsers(
  input: UserInput[],
  connection: ConnectionOrTransaction
): Promise<string[]> {
  const rawInput = input.map((input) => userToRawUser(input));
  const results = await RawUsers(connection).insert(rawInput);
  return results.map((rawEntity) => rawUserToUser(rawEntity));
}
/** Find many Users in the database by a constraint function */
export async function findUsers(
  query: (
    query: Knex.QueryBuilder<RawUser, RawUser[]>
  ) => unknown,
  connection: ConnectionOrTransaction
): Promise<User[]> {
  const queryBuilder = RawUsers(connection);
  await query(queryBuilder);
  const results = await queryBuilder;
  return results.map((rawEntity) => rawUserToUser(rawEntity));
}

/** Find a User in the database or fail */
export async function findUserOrFail(
  query: (
    query: Knex.QueryBuilder<RawUser, RawUser[]>
  ) => unknown,
  connection: ConnectionOrTransaction
): Promise<User> {
  const queryBuilder = RawUsers(connection);
  await query(queryBuilder);
  queryBuilder.limit(1);
  const results = await queryBuilder;
  if (results.length === 0) {
    throw new Error(
      \\"Error! Unable to find User in findUserOrFail call\\"
    );
  }
  return rawUserToUser(results[0]);
}
/** Update a single User */
export async function updateUser(
  entity: UserInput,
  connection: ConnectionOrTransaction
): Promise<void> {
  const { id, ...rawInput } =
    userToRawUser(entity);
  await RawUsers(connection)
    .update(rawInput)
    .where(\\"id\\", \\"=\\", entity.id)
    .limit(1);
}

/** Update one or more Users under specific conditions */
export async function updateUserWhere(
  updatePayload: Partial<UserInput>,
  query: (query: Knex.QueryBuilder<RawUser, number>) => unknown,
  connection: ConnectionOrTransaction
): Promise<void> {
  const rawUpdatePayload = userToRawUser(updatePayload);
  const queryBuilder =
    RawUsers(connection).update(rawUpdatePayload);
  await query(queryBuilder);
  await queryBuilder;
}
/** Deletes a User */
export async function deleteUser(
  entity: User,
  connection: ConnectionOrTransaction
): Promise<number> {
  return RawUsers<number>(connection)
    .where(\\"id\\", \\"=\\", entity.id)
    .delete()
    .limit(1);
}

/** Delete one or more Users under specific conditions */
export async function deleteUserWhere(
  query: (query: Knex.QueryBuilder<RawUser, number>) => unknown,
  connection: ConnectionOrTransaction
): Promise<number> {
  const queryBuilder = RawUsers<number>(connection);
  await query(queryBuilder);
  queryBuilder.delete();
  return queryBuilder;
}
/**
 * Insert a single UserAccount into the database, returning the inserted ID
 */
export async function insertUserAccount(
  input: UserAccountInput,
  connection: ConnectionOrTransaction
): Promise<string> {
  const rawInput = userAccountToRawUserAccount(input);
  const result = await RawUserAccounts(connection).insert(rawInput);
  return rawUserAccountToUserAccount(result[0]);
}

/**
 * Inserts one ore more UserAccounts into the database, returning the inserted IDs
 */
export async function insertUserAccounts(
  input: UserAccountInput[],
  connection: ConnectionOrTransaction
): Promise<string[]> {
  const rawInput = input.map((input) => userAccountToRawUserAccount(input));
  const results = await RawUserAccounts(connection).insert(rawInput);
  return results.map((rawEntity) => rawUserAccountToUserAccount(rawEntity));
}
/** Find many UserAccounts in the database by a constraint function */
export async function findUserAccounts(
  query: (
    query: Knex.QueryBuilder<RawUserAccount, RawUserAccount[]>
  ) => unknown,
  connection: ConnectionOrTransaction
): Promise<UserAccount[]> {
  const queryBuilder = RawUserAccounts(connection);
  await query(queryBuilder);
  const results = await queryBuilder;
  return results.map((rawEntity) => rawUserAccountToUserAccount(rawEntity));
}

/** Find a UserAccount in the database or fail */
export async function findUserAccountOrFail(
  query: (
    query: Knex.QueryBuilder<RawUserAccount, RawUserAccount[]>
  ) => unknown,
  connection: ConnectionOrTransaction
): Promise<UserAccount> {
  const queryBuilder = RawUserAccounts(connection);
  await query(queryBuilder);
  queryBuilder.limit(1);
  const results = await queryBuilder;
  if (results.length === 0) {
    throw new Error(
      \\"Error! Unable to find UserAccount in findUserAccountOrFail call\\"
    );
  }
  return rawUserAccountToUserAccount(results[0]);
}
/** Update a single UserAccount */
export async function updateUserAccount(
  entity: UserAccountInput,
  connection: ConnectionOrTransaction
): Promise<void> {
  const { id, ...rawInput } =
    userAccountToRawUserAccount(entity);
  await RawUserAccounts(connection)
    .update(rawInput)
    .where(\\"id\\", \\"=\\", entity.id)
    .limit(1);
}

/** Update one or more UserAccounts under specific conditions */
export async function updateUserAccountWhere(
  updatePayload: Partial<UserAccountInput>,
  query: (query: Knex.QueryBuilder<RawUserAccount, number>) => unknown,
  connection: ConnectionOrTransaction
): Promise<void> {
  const rawUpdatePayload = userAccountToRawUserAccount(updatePayload);
  const queryBuilder =
    RawUserAccounts(connection).update(rawUpdatePayload);
  await query(queryBuilder);
  await queryBuilder;
}
/** Deletes a UserAccount */
export async function deleteUserAccount(
  entity: UserAccount,
  connection: ConnectionOrTransaction
): Promise<number> {
  return RawUserAccounts<number>(connection)
    .where(\\"id\\", \\"=\\", entity.id)
    .delete()
    .limit(1);
}

/** Delete one or more UserAccounts under specific conditions */
export async function deleteUserAccountWhere(
  query: (query: Knex.QueryBuilder<RawUserAccount, number>) => unknown,
  connection: ConnectionOrTransaction
): Promise<number> {
  const queryBuilder = RawUserAccounts<number>(connection);
  await query(queryBuilder);
  queryBuilder.delete();
  return queryBuilder;
}
",
  "testCode": "",
}
`;

exports[`generates some code: PG flavored 1`] = `
Object {
  "code": "/**
 * Insert a single User into the database, returning the inserted entity
 */
export async function insertUser(
  input: UserInput,
  connection: ConnectionOrTransaction
): Promise<User> {
  const rawInput = userToRawUser(input);
  const result = await RawUsers(connection)
    .insert(rawInput)
    .returning(\\"*\\");
  return rawUserToUser(result[0]);
}

/**
 * Inserts one ore more Users into the database, returning the inserted entities
 */
export async function insertUsers(
  input: UserInput[],
  connection: ConnectionOrTransaction
): Promise<User[]> {
  const rawInput = input.map((input) => userToRawUser(input));
  const results = await RawUsers(connection)
    .insert(rawInput)
    .returning(\\"*\\");
  return results.map((rawEntity) => rawUserToUser(rawEntity));
}
/** Find many Users in the database by a constraint function */
export async function findUsers(
  query: (
    query: Knex.QueryBuilder<RawUser, RawUser[]>
  ) => unknown,
  connection: ConnectionOrTransaction
): Promise<User[]> {
  const queryBuilder = RawUsers(connection);
  await query(queryBuilder);
  const results = await queryBuilder;
  return results.map((rawEntity) => rawUserToUser(rawEntity));
}

/** Find a User in the database or fail */
export async function findUserOrFail(
  query: (
    query: Knex.QueryBuilder<RawUser, RawUser[]>
  ) => unknown,
  connection: ConnectionOrTransaction
): Promise<User> {
  const queryBuilder = RawUsers(connection);
  await query(queryBuilder);
  queryBuilder.limit(1);
  const results = await queryBuilder;
  if (results.length === 0) {
    throw new Error(
      \\"Error! Unable to find User in findUserOrFail call\\"
    );
  }
  return rawUserToUser(results[0]);
}
/** Update a single User */
export async function updateUser(
  entity: UserInput,
  connection: ConnectionOrTransaction
): Promise<void> {
  const { id, ...rawInput } =
    userToRawUser(entity);
  await RawUsers(connection)
    .update(rawInput)
    .where(\\"id\\", \\"=\\", entity.id)
    .limit(1);
}

/** Update one or more Users under specific conditions */
export async function updateUserWhere(
  updatePayload: Partial<UserInput>,
  query: (query: Knex.QueryBuilder<RawUser, number>) => unknown,
  connection: ConnectionOrTransaction
): Promise<void> {
  const rawUpdatePayload = userToRawUser(updatePayload);
  const queryBuilder =
    RawUsers(connection).update(rawUpdatePayload);
  await query(queryBuilder);
  await queryBuilder;
}
/** Deletes a User */
export async function deleteUser(
  entity: User,
  connection: ConnectionOrTransaction
): Promise<number> {
  return RawUsers<number>(connection)
    .where(\\"id\\", \\"=\\", entity.id)
    .delete()
    .limit(1);
}

/** Delete one or more Users under specific conditions */
export async function deleteUserWhere(
  query: (query: Knex.QueryBuilder<RawUser, number>) => unknown,
  connection: ConnectionOrTransaction
): Promise<number> {
  const queryBuilder = RawUsers<number>(connection);
  await query(queryBuilder);
  queryBuilder.delete();
  return queryBuilder;
}
/**
 * Insert a single UserAccount into the database, returning the inserted entity
 */
export async function insertUserAccount(
  input: UserAccountInput,
  connection: ConnectionOrTransaction
): Promise<UserAccount> {
  const rawInput = userAccountToRawUserAccount(input);
  const result = await RawUserAccounts(connection)
    .insert(rawInput)
    .returning(\\"*\\");
  return rawUserAccountToUserAccount(result[0]);
}

/**
 * Inserts one ore more UserAccounts into the database, returning the inserted entities
 */
export async function insertUserAccounts(
  input: UserAccountInput[],
  connection: ConnectionOrTransaction
): Promise<UserAccount[]> {
  const rawInput = input.map((input) => userAccountToRawUserAccount(input));
  const results = await RawUserAccounts(connection)
    .insert(rawInput)
    .returning(\\"*\\");
  return results.map((rawEntity) => rawUserAccountToUserAccount(rawEntity));
}
/** Find many UserAccounts in the database by a constraint function */
export async function findUserAccounts(
  query: (
    query: Knex.QueryBuilder<RawUserAccount, RawUserAccount[]>
  ) => unknown,
  connection: ConnectionOrTransaction
): Promise<UserAccount[]> {
  const queryBuilder = RawUserAccounts(connection);
  await query(queryBuilder);
  const results = await queryBuilder;
  return results.map((rawEntity) => rawUserAccountToUserAccount(rawEntity));
}

/** Find a UserAccount in the database or fail */
export async function findUserAccountOrFail(
  query: (
    query: Knex.QueryBuilder<RawUserAccount, RawUserAccount[]>
  ) => unknown,
  connection: ConnectionOrTransaction
): Promise<UserAccount> {
  const queryBuilder = RawUserAccounts(connection);
  await query(queryBuilder);
  queryBuilder.limit(1);
  const results = await queryBuilder;
  if (results.length === 0) {
    throw new Error(
      \\"Error! Unable to find UserAccount in findUserAccountOrFail call\\"
    );
  }
  return rawUserAccountToUserAccount(results[0]);
}
/** Update a single UserAccount */
export async function updateUserAccount(
  entity: UserAccountInput,
  connection: ConnectionOrTransaction
): Promise<void> {
  const { id, ...rawInput } =
    userAccountToRawUserAccount(entity);
  await RawUserAccounts(connection)
    .update(rawInput)
    .where(\\"id\\", \\"=\\", entity.id)
    .limit(1);
}

/** Update one or more UserAccounts under specific conditions */
export async function updateUserAccountWhere(
  updatePayload: Partial<UserAccountInput>,
  query: (query: Knex.QueryBuilder<RawUserAccount, number>) => unknown,
  connection: ConnectionOrTransaction
): Promise<void> {
  const rawUpdatePayload = userAccountToRawUserAccount(updatePayload);
  const queryBuilder =
    RawUserAccounts(connection).update(rawUpdatePayload);
  await query(queryBuilder);
  await queryBuilder;
}
/** Deletes a UserAccount */
export async function deleteUserAccount(
  entity: UserAccount,
  connection: ConnectionOrTransaction
): Promise<number> {
  return RawUserAccounts<number>(connection)
    .where(\\"id\\", \\"=\\", entity.id)
    .delete()
    .limit(1);
}

/** Delete one or more UserAccounts under specific conditions */
export async function deleteUserAccountWhere(
  query: (query: Knex.QueryBuilder<RawUserAccount, number>) => unknown,
  connection: ConnectionOrTransaction
): Promise<number> {
  const queryBuilder = RawUserAccounts<number>(connection);
  await query(queryBuilder);
  queryBuilder.delete();
  return queryBuilder;
}
",
  "testCode": "",
}
`;
